### ISSUE TITLE
Retreat ignores "unitsNotAllowed"

### ISSUE THREAD

**Issue by alkexr** [triplea-game/triplea#4107]
### Engine version
11665 pre-release
Most likely affects all engine versions

### Bug description
Retreat ignores "unitsNotAllowed" of the territory effect

### How to trigger the bug?
Use e.g. Battle for Arda. Attack a territory from 2 directions:
* from mountains with units that can enter mountains
* from another direction with units that can't enter montains
Then retreat to the mountains.

[retreattomountain.zip](https://github.com/triplea-game/triplea/files/2418876/retreattomountain.zip) (here only rangers should be able to enter mountains)

### Instead of this, what should have happened?
Not sure. General solution would require retreat to multiple territories OR disallowing retreat when no territory is suitable for all retreating units.

**Comment by Cernelius** [comment]
This is a same problem also for movement stack limits.

**Comment by ron-murhammer** [comment]
Yeah, generally retreat locations are missing some verification for some of the more advanced properties. I think it should probably work that you can only retreat to 1 location and only if all units should be allowed to move there.

**Comment by Cernelius** [comment]
Just for at least supporting the matter at hand, that is most likely the right solution. However, solving this issue would be highly complementary with adding some option for a more refined way to handle retreat, that, then, a mapmaker would just assure to be using when having such options in its map. Once you have the ability to specify units for stack limits / canals / territory effects, the ability to retreat all to 1 single of the territories you came from ends up being really to simplistic and limited. Simple example: you have units that cannot go in territory effect A and units that cannot go in territory effect B; they come from territories having such effects, thus nobody can possibly retreat until all the other ones are dead.

**Comment by ron-murhammer** [comment]
True though the only simple solution to that is to have an option where units retreat to where they came from.

**Comment by Cernelius** [comment]
@ron-murhammer That may be a simple solution, but, as @simon33-2 points out, the challenge would be for same units types, coming from different territories, that you would either have to decide which ones are getting killed each combat round or you would need having it randomized (and in this case that should be sorted at the moment you retreat, not each combat round, or you would still need to display it to the user, as that may be important for casualties choices).

We should really open a feature request in forum about this, as this would be a separate feature (either a property or a unit option), but I think the best solution would be to generally allow for units to end combat in the same territory (just like air does), and getting 1 remaining movement for that (no matter how much they used in combat movement), and, in non combat move or at the end of the combat phase, being able to move them (of 1), but only for going to territories whence any of the units that were in there (no matter if any of those surviving) came from. So, I guess, probably better at the end of the combat phase, to not make confusion with non combat movements, since all connections but the ones from which any combat movements were made would need to be temporary made unusable. Practically, exactly like if they were all air units using all movement but 1 to get to the target, but with the additional limit of being limited to use a connection that was used during combat movement by any not-air units (excluding those moved with air transports). Then another property to destroy all your units in hostile territories, after that, to force you to actually retreat all (this could be used also for limited combat rounds, if wanted, to change it into a victory or death timed thing).

**Comment by simon33-2** [comment]
Another possible idea would be to just have the unretreatable units continue to fight, like they do when you have an amphibious assault. What do you think of that idea?

**Comment by VictoryFirst2** [comment]
Tested this with the current engine and it is not a problem anymore. Closing this.
